File: backend/src/modules/gateway/adapters/gpt-codex/builtin-tools.ts
export function normalizeOpenAICodexBuiltinTools(body: Record<string, unknown>): void {
  const tools = Array.isArray(body.tools) ? body.tools : undefined
  if (tools) {
    const normalizedTools = tools.map((tool) => (
      isPlainObject(tool) ? normalizeOpenAICodexBuiltinTool(tool) : tool
    ))
    if (normalizedTools.some((tool, index) => tool !== tools[index])) {
      body.tools = normalizedTools
    }
  }

  const toolChoice = isPlainObject(body.tool_choice) ? body.tool_choice : undefined
  if (toolChoice) {
    const normalizedToolChoice = normalizeOpenAICodexBuiltinTool(toolChoice)
    const toolChoiceTools = Array.isArray(toolChoice.tools) ? toolChoice.tools : undefined
    if (!toolChoiceTools) {
      if (normalizedToolChoice !== toolChoice) {
        body.tool_choice = normalizedToolChoice
      }
      return
    }
    const normalizedTools = toolChoiceTools.map((tool) => (
      isPlainObject(tool) ? normalizeOpenAICodexBuiltinTool(tool) : tool
    ))
    if (
      normalizedToolChoice !== toolChoice
      || normalizedTools.some((tool, index) => tool !== toolChoiceTools[index])
    ) {
      body.tool_choice = {
        ...normalizedToolChoice,
        tools: normalizedTools
      }
    }
  }
}

function normalizeOpenAICodexBuiltinTool(value: Record<string, unknown>): Record<string, unknown> {
  const current = typeof value.type === 'string' ? value.type : ''
  if (current === 'web_search_preview' || current === 'web_search_preview_2025_03_11') {
    return { ...value, type: 'web_search' }
  }
  return value
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

File: backend-go/projects/gateway/internal/gatewaydispatch/builtintools.go
package gatewaydispatch

import "encoding/json"

// Codex builtin tool normalization, migrated from
// adapters/gpt-codex/builtin-tools.ts.

// NormalizeOpenAICodexBuiltinTools mirrors normalizeOpenAICodexBuiltinTools:
// web_search_preview* tool types collapse to web_search.
func NormalizeOpenAICodexBuiltinTools(body map[string]any) {
	if tools, ok := body["tools"].([]any); ok {
		normalizedTools := make([]any, len(tools))
		changed := false
		for index, tool := range tools {
			normalized := normalizeOpenAICodexBuiltinTool(tool)
			normalizedTools[index] = normalized
			if !jsonValueEqual(normalized, tool) {
				changed = true
			}
		}
		if changed {
			body["tools"] = normalizedTools
		}
	}

	if toolChoice, ok := body["tool_choice"].(map[string]any); ok {
		normalizedToolChoice := normalizeOpenAICodexBuiltinTool(toolChoice)
		toolChoiceTools, hasTools := toolChoice["tools"].([]any)
		if !hasTools {
			if !jsonValueEqual(normalizedToolChoice, toolChoice) {
				body["tool_choice"] = normalizedToolChoice
			}
			return
		}
		normalizedTools := make([]any, len(toolChoiceTools))
		toolsChanged := false
		for index, tool := range toolChoiceTools {
			normalized := normalizeOpenAICodexBuiltinTool(tool)
			normalizedTools[index] = normalized
			if !jsonValueEqual(normalized, tool) {
				toolsChanged = true
			}
		}
		if !jsonValueEqual(normalizedToolChoice, toolChoice) || toolsChanged {
			choiceObject, _ := normalizedToolChoice.(map[string]any)
			merged := make(map[string]any, len(choiceObject)+1)
			for key, value := range choiceObject {
				merged[key] = value
			}
			merged["tools"] = normalizedTools
			body["tool_choice"] = merged
		}
	}
}

func normalizeOpenAICodexBuiltinTool(value any) any {
	object, ok := value.(map[string]any)
	if !ok {
		return value
	}
	current, _ := object["type"].(string)
	if current == "web_search_preview" || current == "web_search_preview_2025_03_11" {
		replacement := make(map[string]any, len(object))
		for key, item := range object {
			replacement[key] = item
		}
		replacement["type"] = "web_search"
		return replacement
	}
	return value
}

// jsonValueEqual compares two decoded JSON values structurally.
func jsonValueEqual(left, right any) bool {
	leftRaw, leftErr := json.Marshal(left)
	rightRaw, rightErr := json.Marshal(right)
	if leftErr != nil || rightErr != nil {
		return false
	}
	return string(leftRaw) == string(rightRaw)
}
